object: add support for user s3 key for cephobjectstoreuser (#14001)
rook/rook: Documentation/CRDs/Object-Storage/ceph-object-store-user-crd.md

    * `user-policy`
    * `odic-provider`
    * `ratelimit`

> ### CephObjectStoreUser Reference Secret(optional)

Reviewer: Optional can be implied. 
```suggestion
### CephObjectStoreUser Reference Secret
```

> User can create a k8s secret where S3 key is user defined.

Reviewer: ```suggestion
If a specific user key and secret is desired instead of randomly generated credentials, a specific user key and secret can be specified for an object store user. 
```

> Create or Update the secret with name, `rook-ceph-object-user-<store-name>-<user-name>` in the same namespace of cephobjectUser, where

Reviewer: ```suggestion
Create or update the Kubernetes secret with name, `rook-ceph-object-user-<store-name>-<user-name>` in the same namespace of cephobjectUser, where:

* `store-name`: The object store name in which the user will be created. This matches the name of the objectstore CRD.
* `user-name`: The metadata name of the cephObjectStoreUser

```console
kubectl create -f
apiVersion: v1
kind: Secret
metadata:
  name: rook-ceph-object-user-my-store-my-user
  namespace: rook-ceph
data:
  AccessKey: ***
  SecretKey: ***
```

> * `AccessKey` and `SecretKey`: S3 access and secret keys

Reviewer: I'd suggest no need for this, it seems obvious
```suggestion
```

    * `user-policy`
    * `odic-provider`
    * `ratelimit`

### CephObjectStoreUser Reference Secret

If a specific user key and secret is desired instead of randomly generated credentials, a specific user key and secret can be specified for an object store user.

Create or update the Kubernetes secret with name, `rook-ceph-object-user-<store-name>-<user-name>` in the same namespace of cephobjectUser, where:

* `store-name`: The object store name in which the user will be created. This matches the name of the objectstore CRD.
* `user-name`: The metadata name of the cephObjectStoreUser

> 

Author: something like this

```console
kubectl create -f
apiVersion: v1
kind: Secret
metadata:
  name: rook-ceph-object-user-my-store-my-user
  namespace: rook-ceph
  annotations:

>     rook.io/source-of-truth: secret

Author: like this

    * `user-policy`
    * `odic-provider`
    * `ratelimit`

### CephObjectStoreUser Reference Secret

> If a specific user key and secret is desired instead of randomly generated credentials, a specific user key and secret can be specified for an object store user.

Author: Should I add in this same document or  create a separate file for this?

Create or update the Kubernetes secret with name, `rook-ceph-object-user-<store-name>-<user-name>` in the same namespace of cephobjectUser, where:

* `store-name`: The object store name in which the user will be created. This matches the name of the objectstore CRD.
* `user-name`: The metadata name of the cephObjectStoreUser

#### Secret details:

i) The annotations as `rook.io/source-of-truth: secret` specified and type as `type: "kubernetes.io/rook"`.

ii) (optional) The array of `SecretKeys` which contains all the truted access and secret keys. If any key is present in the ceph user and not present in the `SecretKeys`, it will be removed from the ceph user too.

If `SecretKeys` is left empty, it will be auto updated by the latest AccessKey and SecretKey and will remove all the other keys from the ceph user.

```console
`SecretKeys`
[{"user":"","access_key":"IE58RNT71Y2F1EQE80RA","secret_key":"cULyMz5dCpX18dPsJhpIKay7vcDNRNJWJPu8VqUA","UID":"","SubUser":"","KeyType":"","GenerateKey":null},

> {"user":"","access_key":"IE58RNT71Y2F1EQE80RZ","secret_key":"cULyMz5dCpX18dPsJhpIKay7vcDNRNJWJPu8VqUE","UID":"","SubUser":"","KeyType":"","GenerateKey":null}]

Author: We also need UID so updating with that

> {"user":"","access_key":"IE58RNT71Y2F1EQE80RZ","secret_key":"cULyMz5dCpX18dPsJhpIKay7vcDNRNJWJPu8VqUE","UID":"","SubUser":"","KeyType":"","GenerateKey":null}]

Author: Or maybe I can update to object-user name for all the users

Convert it to base64:
W3sidXNlciI6IiIsImFjY2Vzc19rZXkiOiJJRTU4Uk5UNzFZMkYxRVFFODBSQSIsInNlY3JldF9rZXkiOiJjVUx5TXo1ZENwWDE4ZFBzSmhwSUtheTd2Y0ROUk5KV0pQdThWcVVBIiwiVUlEIjoiIiwiU3ViVXNlciI6IiIsIktleVR5cGUiOiIiLCJHZW5lcmF0ZUtleSI6bnVsbH0seyJ1c2VyIjoiIiwiYWNjZXNzX2tleSI6IklFNThSTlQ3MVkyRjFFUUU4MFJaIiwic2VjcmV0X2tleSI6ImNVTHlNejVkQ3BYMThkUHNKaHBJS2F5N3ZjRE5STkpXSlB1OFZxVUUiLCJVSUQiOiIiLCJTdWJVc2VyIjoiIiwiS2V5VHlwZSI6IiIsIkdlbmVyYXRlS2V5IjpudWxsfV0=
```

!!! note
    All the data values should be converted to base64 format and then updated.

Example Secret:
```console
kubectl create -f
apiVersion: v1
kind: Secret
metadata:
  name: rook-ceph-object-user-my-store-my-user
  namespace: rook-ceph
  annotations:
    rook.io/source-of-truth: secret
data:
  AccessKey: ***
  SecretKey: ***

>   Endpoint: ***

Author: Or maybe I can update to object-user name for all the users

    * `user-policy`
    * `odic-provider`
    * `ratelimit`

### CephObjectStoreUser Reference Secret

A Kubernetes secret is created by Rook with the name, `rook-ceph-object-user-<store-name>-<user-name>` in the same namespace of cephobjectUser, where:

* `store-name`: The object store name in which the user will be created. This matches the name of the objectstore CRD.
* `user-name`: The metadata name of the cephObjectStoreUser

This secret can be used to configure there applications.
Secret contains,

> `AccessKey`, `Credentials`, `Endpoint`, `Endpoint`.

Reviewer 2: Probably not necessary since the below block says what the secret contains.

    * `user-policy`
    * `odic-provider`
    * `ratelimit`

### CephObjectStoreUser Reference Secret

A Kubernetes secret is created by Rook with the name, `rook-ceph-object-user-<store-name>-<user-name>` in the same namespace of cephobjectUser, where:

* `store-name`: The object store name in which the user will be created. This matches the name of the objectstore CRD.

> * `user-name`: The metadata name of the cephObjectStoreUser

Reviewer 2: CephObjectStoreUser is the name of a Rook-defined custom resource and should always be documented in UpperCamelCase format. lowerCamelCase should be used for the names of fields within.

```suggestion
* `user-name`: The metadata name of the CephObjectStoreUser
```

Application pods can use the secret to configure access to the object store.

```yaml
apiVersion: v1
data:
  AccessKey: ***       # [1]
  Credentials: ***     # [2]
  Endpoint: ***        # [3]
  SecretKey: ***       # [4]
kind: Secret
...
...
type: kubernetes.io/rook

```

1. `AccessKey`: User access key for ceph S3.
2. `SecretKey`: User secret key for ceph S3.

> 3. `Credentials`: Comma separated values of all access and secret keys.

Reviewer 2: Let's add an explanation of the data format so users know how to interact with it.

4. `Endpoint`: Endpoint for ceph S3.

#### User-Specified Reference Secret

> AccessKey, SecretKey, and Credentials can be specified before CephObjectStore creation or modified afterwards to support user-specified keys and key rotation. Endpoint cannot be overridden by users and will always be updated by Rook to match the latest CephObjectStore information.

Reviewer 2: As mentioned, I am concerned about how users reading the doc may misconstrue the "user" in "user-specified" with the "user" in "CephObjectStoreUser." Let us try to find clearer wording for "user-specified" that doesn't have the potential to confuse readers.

```suggestion
AccessKey, SecretKey, and Credentials can be specified before CephObjectStore creation or modified afterwards to support custom access credentials and key rotation. Endpoint cannot be overridden by users and will always be updated by Rook to match the latest CephObjectStore information.
```
